### ISSUE TITLE
dmypy: error: unrecognized arguments: --config-file pyproject.toml

### ISSUE THREAD

**Issue by petergaultney** [python/mypy#17624]
**Bug Report**

**To Reproduce**

Run the `dmypy` daemon as `dmypy run -- . --config-file pyproject.toml`

**Expected Behavior**

`dmypy` should respect [the part of its own documentation](https://mypy.readthedocs.io/en/stable/mypy_daemon.html#basic-usage) that states "You can use almost arbitrary mypy flags after --"

**Actual Behavior**

This fails with

```
usage: dmypy [-h] [--status-file STATUS_FILE] [-V] {start,restart,status,stop,kill,check,run,recheck,suggest,inspect,hang,daemon,help} ...
dmypy: error: unrecognized arguments: --config-file pyproject.toml
```

And if I remove that part and run only `dmypy run -- .`, it runs fine, except it ignores the mypy config in that file.
**Your Environment**

- Mypy version used: 1.10.0, 1.10.1, 1.11.0 (i tested all three)
- Mypy command-line flags: (as above)
- Mypy configuration options from `mypy.ini` (and other config files): N/A, see above
- Python version used: 3.8, 3.10

**Comment by ricwo** [comment]
Bumping this issue. Also happens in 1.11.2 (Python 3.12).

**Comment by kcornw** [comment]
Hello! I failed to reproduce it in mypy 1.13.0 (python 3.12.3).

```
$ dmypy run -- . --config-file custom.toml
Success: no issues found in 1 source file
```

Is it already fixed?

**Comment by brianschubert** [comment]
I'm having trouble reproducing this issue on master or any of the affected versions above. Some more details about exactly how dmypy is being run would be helpful (e.g. what environment/shell, what sort of terminal or editor you're using to type the command). 

My slight suspicion is that the environment that dmypy is being run in isn't passing the `--` sequence properly. That would result in the sort of error message that's being seen above. For example, if `--` is being passed with some extra embedded whitespace, or is being replaced by some unicode character:
```shell
# good, plain old '--'
$ dmypy run -- . --config-file pyproject.toml       
Success: no issues found in 1 source file

# bad: extra whitespace
$ dmypy run $'‐‐\n' . --config-file pyproject.toml  
usage: dmypy [-h] [--status-file STATUS_FILE] [-V] {start,restart,status,stop,kill,check,run,recheck,suggest,inspect,hang,daemon,help} ...
dmypy: error: unrecognized arguments: --config-file pyproject.toml

# bad: replaced by em dash
$ dmypy run — . --config-file pyproject.toml        
usage: dmypy [-h] [--status-file STATUS_FILE] [-V] {start,restart,status,stop,kill,check,run,recheck,suggest,inspect,hang,daemon,help} ...
dmypy: error: unrecognized arguments: --config-file pyproject.toml
```

**Comment by petergaultney** [comment]
I can still reproduce this trivially in mypy 1.13.

I copy-pasted the "plain old '--'" example @brianschubert provided above and that made no difference to me (I wasn't typing em or en dashes in my original example). I'm not sure what could be inserting some kind of extra whitespace into this command without my knowledge.

I'll try some other shells to see if anything changes the behavior.

**Comment by petergaultney** [comment]
Shells make no difference. However ---

It would seem that `poetry run` is somehow related to this problem. If I launch a poetry subshell, `dmypy` seems to work. If I run `dmypy` by using the path to `.venv/bin/dmypy` directly, it also works fine. 

Somehow, this doesn't shock me, even though you'd think that `poetry run` would be a pretty well-tested command by this point. 

I don't really understand the error, though. Why is `dmypy` itself telling me that it doesn't recognize the arguments?  If dmypy is successfully launching and successfully parsing the arguments, shouldn't it be able to provide a better error message that this? I see no indication of extra whitespace (or anything else) in its error message.

**Comment by wyattscarpenter** [comment]
@petergaultney I was able to reproduce this issue. My hunch is that you are running

```
poetry run dmypy run -- . --config-file pyproject.toml
```

(which gets me the same error, `dmypy: error: unrecognized arguments: --config-file pyproject.toml`)

but you need to be running

```
poetry run -- dmypy run -- . --config-file pyproject.toml
```

The problem is that poetry run can also accept a -- argument, and will just take one from anywhere, as you can see from

```
poetry run echo run -- . --config-file pyproject.toml
```
which produces:

```
run . --config-file pyproject.toml
```

so dmypy is not getting a --, and isn't passing the arguments along to the rest of mypy.

This is, of late, a known... bug? feature?... of poetry, it turns out: https://github.com/python-poetry/poetry/issues/10204

**Comment by petergaultney** [comment]
yes, that's right. it's still confusing that the issue manifests as a dmypy error - plausibly a good thing for dmypy to do would be to indicate clearly that it received the argument as an argument _to dmypy_ - the error message is on the terse side, and it seems likely that (poetry aside) a common user case here would be passing the argument to the wrong 'part' of mypy - e.g. passing an argument that would be recognized by mypy itself but not by dmypy. 

we, of course, moved on to `uv` anyway, so this bug no longer affects us. I will just close this issue, though I would still be interested in discussion about whether there's a way that dmypy could provide better error messages - perhaps simply by echoing the full string that it received so that users might spot the fact that the `--` had been 'swallowed' upstream
